### ISSUE TITLE
Build Question

### ISSUE THREAD

**Issue by mrberman87** [CopernicaMarketingSoftware/AMQP-CPP#127]
I am having issues compiling strait from the examples in the the README.
I am running this in a VM on VirtualBox.  I am running Ubuntu 16.04.2 LTS with all updates.  I am running AMQP-CPP version 2.7 built from source.

Below is my source code, something simple just to get compilation at this point, but I am getting an undefined reference in TcpConnection.  The compilation output is below the source.

After compiling, I have run ldconfig, and the output of 'ldconfig -p' | grep amqp' shows libamqpcpp.so

`
#include <amqpcpp.h>
#include <iostream>

struct MyTcpHandler : public AMQP::TcpHandler
{
    virtual void onConnected(AMQP::TcpConnection *connection)
    {
        std::cout << "MyTcpHandler: " << __FUNCTION__ << std::endl;
    }

    virtual void onError(AMQP::TcpConnection *connection, const char *message)
    {
        std::cout << "MyTcpHandler: " << __FUNCTION__ << std::endl;
        std::cout << std::string(message) << std::endl;
    }

    virtual void onClosed(AMQP::TcpConnection *connection)
    {
        std::cout << "MyTcpHandler: " << __FUNCTION__ << std::endl;
    }

    virtual void monitor(AMQP::TcpConnection *connection, int fd, int flags)
    {
        std::cout << "MyTcpHandler: " << __FUNCTION__ << std::endl;
    }
};

int main()
{
    std::cout << "Hello from " << __FUNCTION__ << std::endl;

    // create an instance of your own tcp handler
    MyTcpHandler myHandler;

    AMQP::Address address("amqp://localhost/");

    // create a AMQP connection object
    AMQP::TcpConnection connection(&myHandler, address);

    // and create a channel
    //AMQP::TcpChannel channel(&connection);

    // use the channel object to call the AMQP method you like
    //channel.declareExchange("", AMQP::fanout);
    //channel.declareQueue("hello");
    //channel.bindQueue("", "hello", "hello");

    return 0;
}
`

mkdir -p build
g++ -Wall -Wno-deprecated-declarations -g -std=c++14 -pthread -I/usr/include -L/usr/lib -lamqpcpp -Isrc -c src/run.cpp -o build/run.o
g++ -Wall -Wno-deprecated-declarations -g -std=c++14 -pthread -I/usr/include -L/usr/lib -lamqpcpp -Isrc -o run.out build/run.o
build/run.o: In function `main':
/home/work/Devel/playground/rabbitmq/amqp-cpp-learning/src/run.cpp:38: undefined reference to `AMQP::TcpConnection::TcpConnection(AMQP::TcpHandler*, AMQP::Address const&)'
build/run.o: In function `AMQP::Connection::~Connection()':
/usr/include/amqpcpp/connection.h:70: undefined reference to `AMQP::ConnectionImpl::~ConnectionImpl()'
build/run.o: In function `AMQP::TcpConnection::~TcpConnection()':
/usr/include/amqpcpp/tcpconnection.h:118: undefined reference to `vtable for AMQP::TcpConnection'
/usr/include/amqpcpp/tcpconnection.h:118: undefined reference to `vtable for AMQP::TcpConnection'
/usr/include/amqpcpp/tcpconnection.h:118: undefined reference to `AMQP::Watchable::~Watchable()'
collect2: error: ld returned 1 exit status
Makefile:19: recipe for target 'run.out' failed
make: *** [run.out] Error 1

**Comment by wshn13** [comment]
Same problem ...

**Comment by adibhannachi** [comment]
Same problem ...

**Comment by georgikoyrushki95** [comment]
I got the same problem... I'm pretty sure the library is not built properly. I can see that the TcpConnection class is inside the src directory. Did anyone manage to resolve this?

**Comment by EmielBruijntjes** [comment]
I cannot reproduce this problem. Please provide the exact steps that you took all the way from downloading AMQP up to running your test. 

I copied your code to the file "test.cpp" in my home directory, and fixed some small small issues that were probably caused by the github editor (like underscores that were missing). I then did the following things to install AMQP-CPP, compile your example program and get it up and running:

```
git clone git@github.com:CopernicaMarketingSoftware/AMQP-CPP.git
cd AMQP-CPP
make
sudo make install
cd ~
g++ -std=c++11 test.cpp -lamqpcpp -lpthread
./a.out
```

**Comment by zerodefect** [comment]
I have seen this when a project is mixing libraries that do/don't use position independent code (gcc's -fPIC) command.  

In cmake, this can be resolved using:

`cmake -DCMAKE_BUILD_TYPE=DEBUG -DCMAKE_POSITION_INDEPENDENT_CODE=ON <path_to_root>`

This is only coming from experience; I've not seen anything in the log to indicate this is the problem.

**Comment by gchaviaras** [comment]
@EmielBruijntjes I had the same issue. Apparently I solve the issue by compiling the release package of the library located here https://github.com/CopernicaMarketingSoftware/AMQP-CPP/releases .
Cloning the repo gave me the same result as @mrberman87 , could you please review that issue?
Thanks

**Comment by EmielBruijntjes** [comment]
I am still waiting on the *exact styeps* to reproduce this issue.

**Comment by zerodefect** [comment]
@gchaviaras, can you try change '__FUNCTION__ ' to '__func__' and let us know if it then works?

**Comment by gchaviaras** [comment]
Cant reproduce right now, I will try to deploy a fresh install

**Comment by EmielBruijntjes** [comment]
I am still waiting for the **exact steps** to reproduce this issue.

**Comment by chaithubk** [comment]
Hi,

Am pretty new to windows and RabbitMQ. Sorry for the lengthy write-up, it's an attempt to explain my build issue clearly. Thanks for reading.

I am developing a AMQP client in C++. For that I thought I will use AMQP-CPP library and start developing the client. I took the latest repo via "Download Zip" option in the git hub. I am able to build the library successfully on Windows using cmake. This is my cmake output.

C:\GIT\AMQP-CPP-master>mkdir build & cd build
C:\GIT\AMQP-CPP-master\build>cmake ..
-- Building for: Visual Studio 14 2015
-- Selecting Windows SDK version  to target Windows 10.0.16299.
-- The C compiler identification is MSVC 19.0.24215.1
-- The CXX compiler identification is MSVC 19.0.24215.1
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/cl.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: C:/GIT/AMQP-CPP-master/build

C:\GIT\AMQP-CPP-master\build>

I have created a default Win32 console application in VS 2015. And my main CPP file looks like this

#include "stdafx.h"
#ifdef NOMINMAX
#include <windows.h>
#else
#define NOMINMAX
#include <windows.h>
#undef NOMINMAX
#endif
#include <amqpcpp.h>

int main()
{
	return 0;
}

When I try to build, am seeing this issues and my solution is not building fine.

1>------ Build started: Project: amqp_cpp_client1, Configuration: Debug Win32 ------
1>  stdafx.cpp
1>  amqp_cpp_client1.cpp
1>c:\program files (x86)\windows kits\8.1\include\shared\ws2def.h(100): warning C4005: 'AF_IPX': macro redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(452): note: see previous definition of 'AF_IPX'
1>c:\program files (x86)\windows kits\8.1\include\shared\ws2def.h(140): warning C4005: 'AF_MAX': macro redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(471): note: see previous definition of 'AF_MAX'
1>c:\program files (x86)\windows kits\8.1\include\shared\ws2def.h(177): warning C4005: 'SO_DONTLINGER': macro redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(394): note: see previous definition of 'SO_DONTLINGER'
1>c:\program files (x86)\windows kits\8.1\include\shared\ws2def.h(221): error C2011: 'sockaddr': 'struct' type redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(1002): note: see declaration of 'sockaddr'
1>c:\program files (x86)\windows kits\8.1\include\shared\ws2def.h(421): error C2059: syntax error: 'constant'
1>c:\program files (x86)\windows kits\8.1\include\shared\ws2def.h(421): error C3805: 'constant': unexpected token, expected either '}' or a ','
1>c:\program files (x86)\windows kits\8.1\include\shared\ws2def.h(556): warning C4005: 'IN_CLASSA': macro redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(279): note: see previous definition of 'IN_CLASSA'
1>c:\program files (x86)\windows kits\8.1\include\shared\ws2def.h(562): warning C4005: 'IN_CLASSB': macro redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(285): note: see previous definition of 'IN_CLASSB'
1>c:\program files (x86)\windows kits\8.1\include\shared\ws2def.h(568): warning C4005: 'IN_CLASSC': macro redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(291): note: see previous definition of 'IN_CLASSC'
1>c:\program files (x86)\windows kits\8.1\include\shared\ws2def.h(579): warning C4005: 'INADDR_ANY': macro redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(296): note: see previous definition of 'INADDR_ANY'
1>c:\program files (x86)\windows kits\8.1\include\shared\ws2def.h(581): warning C4005: 'INADDR_BROADCAST': macro redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(298): note: see previous definition of 'INADDR_BROADCAST'
1>c:\program files (x86)\windows kits\8.1\include\shared\ws2def.h(615): error C2011: 'sockaddr_in': 'struct' type redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(1006): note: see declaration of 'sockaddr_in'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(136): error C2011: 'fd_set': 'struct' type redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(1014): note: see declaration of 'fd_set'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(156): warning C4005: 'FD_CLR': macro redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(89): note: see previous definition of 'FD_CLR'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(171): warning C4005: 'FD_SET': macro redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(94): note: see previous definition of 'FD_SET'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(180): error C2011: 'timeval': 'struct' type redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(1030): note: see declaration of 'timeval'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(236): error C2011: 'hostent': 'struct' type redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(1018): note: see declaration of 'hostent'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(249): error C2011: 'netent': 'struct' type redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(172): note: see declaration of 'netent'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(256): error C2011: 'servent': 'struct' type redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(1022): note: see declaration of 'servent'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(268): error C2011: 'protoent': 'struct' type redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(1026): note: see declaration of 'protoent'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(364): error C2011: 'WSAData': 'struct' type redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(314): note: see declaration of 'WSAData'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(462): error C2011: 'sockproto': 'struct' type redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(486): note: see declaration of 'sockproto'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(504): error C2011: 'linger': 'struct' type redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(1010): note: see declaration of 'linger'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(517): warning C4005: 'SOMAXCONN': macro redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(536): note: see previous definition of 'SOMAXCONN'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(551): warning C4005: 'FD_READ': macro redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(554): note: see previous definition of 'FD_READ'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(554): warning C4005: 'FD_WRITE': macro redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(555): note: see previous definition of 'FD_WRITE'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(557): warning C4005: 'FD_OOB': macro redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(556): note: see previous definition of 'FD_OOB'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(560): warning C4005: 'FD_ACCEPT': macro redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(557): note: see previous definition of 'FD_ACCEPT'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(563): warning C4005: 'FD_CONNECT': macro redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(558): note: see previous definition of 'FD_CONNECT'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(566): warning C4005: 'FD_CLOSE': macro redefinition
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(559): note: see previous definition of 'FD_CLOSE'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(1646): error C2375: 'accept': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(734): note: see declaration of 'accept'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(1668): error C2375: 'bind': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(739): note: see declaration of 'bind'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(1689): error C2375: 'closesocket': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(744): note: see declaration of 'closesocket'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(1706): error C2375: 'connect': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(746): note: see declaration of 'connect'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(1727): error C2375: 'ioctlsocket': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(751): note: see declaration of 'ioctlsocket'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(1748): error C2375: 'getpeername': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(756): note: see declaration of 'getpeername'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(1769): error C2375: 'getsockname': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(761): note: see declaration of 'getsockname'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(1790): error C2375: 'getsockopt': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(766): note: see declaration of 'getsockopt'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(1815): error C2375: 'htonl': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(773): note: see declaration of 'htonl'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(1832): error C2375: 'htons': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(775): note: see declaration of 'htons'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(1850): error C2375: 'inet_addr': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(777): note: see declaration of 'inet_addr'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(1868): error C2375: 'inet_ntoa': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(779): note: see declaration of 'inet_ntoa'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(1968): error C2375: 'listen': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(781): note: see declaration of 'listen'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(1987): error C2375: 'ntohl': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(785): note: see declaration of 'ntohl'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2004): error C2375: 'ntohs': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(787): note: see declaration of 'ntohs'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2021): error C2375: 'recv': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(789): note: see declaration of 'recv'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2044): error C2375: 'recvfrom': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(795): note: see declaration of 'recvfrom'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2071): error C2375: 'select': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(803): note: see declaration of 'select'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2096): error C2375: 'send': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(810): note: see declaration of 'send'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2119): error C2375: 'sendto': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(816): note: see declaration of 'sendto'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2146): error C2375: 'setsockopt': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(824): note: see declaration of 'setsockopt'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2171): error C2375: 'shutdown': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(831): note: see declaration of 'shutdown'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2191): error C2375: 'socket': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(835): note: see declaration of 'socket'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2216): error C2375: 'gethostbyaddr': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(842): note: see declaration of 'gethostbyaddr'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2238): error C2375: 'gethostbyname': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(847): note: see declaration of 'gethostbyname'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2255): error C2375: 'gethostname': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(849): note: see declaration of 'gethostname'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2295): error C2375: 'getservbyport': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(853): note: see declaration of 'getservbyport'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2314): error C2375: 'getservbyname': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(857): note: see declaration of 'getservbyname'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2333): error C2375: 'getprotobynumber': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(861): note: see declaration of 'getprotobynumber'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2350): error C2375: 'getprotobyname': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(863): note: see declaration of 'getprotobyname'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2370): error C2375: 'WSAStartup': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(867): note: see declaration of 'WSAStartup'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2390): error C2375: 'WSACleanup': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(871): note: see declaration of 'WSACleanup'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2407): error C2375: 'WSASetLastError': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(873): note: see declaration of 'WSASetLastError'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2424): error C2375: 'WSAGetLastError': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(875): note: see declaration of 'WSAGetLastError'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2445): error C2375: 'WSAIsBlocking': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(877): note: see declaration of 'WSAIsBlocking'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2463): error C2375: 'WSAUnhookBlockingHook': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(879): note: see declaration of 'WSAUnhookBlockingHook'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2481): error C2375: 'WSASetBlockingHook': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(881): note: see declaration of 'WSASetBlockingHook'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2499): error C2375: 'WSACancelBlockingCall': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(883): note: see declaration of 'WSACancelBlockingCall'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2517): error C2375: 'WSAAsyncGetServByName': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(885): note: see declaration of 'WSAAsyncGetServByName'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2545): error C2375: 'WSAAsyncGetServByPort': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(893): note: see declaration of 'WSAAsyncGetServByPort'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2573): error C2375: 'WSAAsyncGetProtoByName': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(901): note: see declaration of 'WSAAsyncGetProtoByName'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2599): error C2375: 'WSAAsyncGetProtoByNumber': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(908): note: see declaration of 'WSAAsyncGetProtoByNumber'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2625): error C2375: 'WSAAsyncGetHostByName': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(915): note: see declaration of 'WSAAsyncGetHostByName'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2651): error C2375: 'WSAAsyncGetHostByAddr': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(922): note: see declaration of 'WSAAsyncGetHostByAddr'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2681): error C2375: 'WSACancelAsyncRequest': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(931): note: see declaration of 'WSACancelAsyncRequest'
1>c:\program files (x86)\windows kits\8.1\include\um\winsock2.h(2699): error C2375: 'WSAAsyncSelect': redefinition; different linkage
1>  c:\program files (x86)\windows kits\8.1\include\um\winsock.h(933): note: see declaration of 'WSAAsyncSelect'
1>c:\git\amqp-cpp-master\include\amqpcpp\outbuffer.h(80): error C2440: 'initializing': cannot convert from 'u_short (__stdcall *)(u_short)' to 'uint16_t'
1>  c:\git\amqp-cpp-master\include\amqpcpp\outbuffer.h(80): note: There is no context in which this conversion is possible
1>c:\git\amqp-cpp-master\include\amqpcpp\outbuffer.h(93): error C2440: 'initializing': cannot convert from 'u_long (__stdcall *)(u_long)' to 'uint32_t'
1>  c:\git\amqp-cpp-master\include\amqpcpp\outbuffer.h(93): note: There is no context in which this conversion is possible
1>c:\git\amqp-cpp-master\include\amqpcpp\outbuffer.h(106): error C2568: '==': unable to resolve function overload
1>  c:\git\amqp-cpp-master\include\amqpcpp\outbuffer.h(106): note: could be 'u_long htonl(u_long)'
1>c:\git\amqp-cpp-master\include\amqpcpp\outbuffer.h(106): error C2568: '|': unable to resolve function overload
1>  c:\git\amqp-cpp-master\include\amqpcpp\outbuffer.h(106): note: could be 'u_long htonl(u_long)'
1>c:\git\amqp-cpp-master\include\amqpcpp\outbuffer.h(129): error C2440: 'initializing': cannot convert from 'u_short (__stdcall *)(u_short)' to 'int16_t'
1>  c:\git\amqp-cpp-master\include\amqpcpp\outbuffer.h(129): note: There is no context in which this conversion is possible
1>c:\git\amqp-cpp-master\include\amqpcpp\outbuffer.h(142): error C2440: 'initializing': cannot convert from 'u_long (__stdcall *)(u_long)' to 'int32_t'
1>  c:\git\amqp-cpp-master\include\amqpcpp\outbuffer.h(142): note: There is no context in which this conversion is possible
1>c:\git\amqp-cpp-master\include\amqpcpp\outbuffer.h(155): error C2568: '==': unable to resolve function overload
1>  c:\git\amqp-cpp-master\include\amqpcpp\outbuffer.h(155): note: could be 'u_long htonl(u_long)'
1>c:\git\amqp-cpp-master\include\amqpcpp\outbuffer.h(155): error C2568: '|': unable to resolve function overload
1>  c:\git\amqp-cpp-master\include\amqpcpp\outbuffer.h(155): note: could be 'u_long htonl(u_long)'
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========

Please help me how to write a c++ client program using the AMQP-CPP library.

**Comment by zerodefect** [comment]
I suspect some of what you are seeing is caused by [this](https://stackoverflow.com/questions/1372480/c-redefinition-header-files-winsock2-h).  I almost feel that this should be in a separate ticket because up until this point all failed builds have been Linux based. I'll let admins use their discretion :)

**Comment by chaithubk** [comment]
@zerodefect 👍  Adding `#define _WINSOCKAPI_ `  resolved the errors and my application built fine. thank you.

**Comment by chichivica** [comment]
just got exactly the same issue as @mrberman87 but successfully resolved that. My problems was I build AMQP-CPP with `cmake` but I didn't installed `libssl-dev`. 
I installed that lib and built with `make` and issue gone.

**Comment by alheio** [comment]
Got exactly the same problem as mrberman87.

kubuntu 16.04
cmake version 3.5.1
make version GNU Make 4.1
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
amqpcpp versions master and tag v3.2.0 (did not check older versions)

This is a strange behavior of linker and add_sources macro. I fixed that with usage of cmake aux_source_directory command instead of add_sources macro.

[patch.txt](https://github.com/CopernicaMarketingSoftware/AMQP-CPP/files/2533832/aux_source_directory.txt)

**Comment by sclaggett** [comment]
This is an old issue, but I ran into it yesterday and wanted to leave a comment in case someone encounters the same problem.

In my case, I was trying to statically link `libamqpcpp.a` on Ubuntu into a shared object that will be dynamically loaded by MySQL as a plugin. The root problem is that shared libraries require position independent code but the `libamqpcpp.a` binary is not built this way by default. The errors generated while building the shared object were quite vague which made diagnosing the issue difficult. The solution is what **zerodefect** mentioned above: pass the `-DCMAKE_POSITION_INDEPENDENT_CODE=ON` flag to cmake when building the library.

**Comment by EmielBruijntjes** [comment]
I did not even know that it was possible to create a *.so file based on *.a files. What flags and arguments do you pass to the compiler for that?

**Comment by sclaggett** [comment]
This is the first time I've had to statically link into a shared object on Linux so it was a learning experience. The plugin is simple and is being manually built like so:

```
g++ -Wall -fPIC -c RabbitHandler.cpp -o RabbitHandler.o
g++ -Wall -fPIC -z defs -shared -DMYSQL_DYNAMIC_PLUGIN -I/usr/include/mysql \
  opsnotify.cc RabbitHandler.o /usr/local/lib/libamqpcpp.a -pthread -ldl \
  -o libopsnotify.so
```
